src/transformStateWithClones.js

```diff
@@ -6,8 +6,37 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let copyState = { ...state };
+
+  actions.forEach((action) => {
+    let newState = { ...copyState };
+
+    switch (action.type) {
+      case 'clear':
+        Object.keys(newState).forEach((key) => {
+          delete newState[key];
+        });
+        break;
+      case 'addProperties':
+        newState = { ...newState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => {
+          delete newState[key];
+        });
```

Comment:
Ensure that `action.keysToRemove` is an array. Adding a check or validation for this will prevent potential runtime errors if `keysToRemove` is not provided or is not an array.